src/scaffoldmaker/utils/tubemesh.py

```diff
@@ -282,70 +282,77 @@ def getCoordinatesFromInner(xInner, d1Inner, d2Inner, d3Inner,
-                                                    vector.normalise(d3Inner[n]), 0.0))
+                    interp.getCubicHermiteCurvature(xSurf[n - elementsCountAround], d2Surf[n - elementsCountAround],
+                                                    xSurf[n], d2Surf[n], vector.normalise(d3Surf[n]), 1.0) +
+                    interp.getCubicHermiteCurvature(xSurf[n], d2Surf[n],
+                                                    xSurf[n + elementsCountAround], d2Surf[n + elementsCountAround],
+                                                    vector.normalise(d3Surf[n]), 0.0))
             curvatureAlong.append(curvature)
 
         for n3 in range(elementsCountThroughWall + 1):
             xi3 = xi3List[n3] if relativeThicknessList else 1.0/elementsCountThroughWall * n3
             for n1 in range(elementsCountAround):
                 n = n2*elementsCountAround + n1
-                norm = d3Inner[n]
-                innerx = xInner[n]
-                outerx = xOuter[n]
-                dWall = [wallThickness*c for c in norm]
+                norm = d3Surf[n]
+                surfx = xSurf[n]
+                extrudedx = xExtrudedSurf[n]
                 # x
-                x = interp.interpolateCubicHermite(innerx, dWall, outerx, dWall, xi3)
+                dWall = [wallThickness * c for c in norm]
+                if outward:
+                    x = interp.interpolateCubicHermite(surfx, dWall, extrudedx, dWall, xi3)
+                else:
+                    x = interp.interpolateCubicHermite(extrudedx, dWall, surfx, dWall, xi3)
                 xList.append(x)
 
                 # dx_ds1
-                factor = 1.0 + wallThickness*xi3 * curvatureAroundInner[n]
-                d1 = [ factor*c for c in d1Inner[n]]
+                factor = 1.0 + wallThickness * xi3 * curvatureAroundSurf[n]
+                d1 = [factor*c for c in d1Surf[n]]
                 d1List.append(d1)
 
                 # dx_ds2
                 curvature = curvatureAlong[n]
-                distance = vector.magnitude([x[i] - xInner[n][i] for i in range(3)])
-                factor = 1.0 - curvature*distance
-                d2 = [ factor*c for c in d2Inner[n]]
+                distance = vector.magnitude([x[i] - xSurf[n][i] for i in range(3)])
+                if outward:
+                    factor = 1.0 - curvature * distance
+                else:
+                    factor = 1.0 + curvature * distance
+                d2 = [factor*c for c in d2Surf[n]]
                 d2List.append(d2)
                 curvatureList.append(curvature)
 
                 #dx_ds3
-                d3 = [c * wallThickness * (relativeThicknessList[n3] if relativeThicknessList else 1.0/elementsCountThroughWall) for c in norm]
+                d3 = [c * -wallOutwardDisplacement * (relativeThicknessList[n3] if relativeThicknessList else 1.0 / elementsCountThroughWall) for c in norm]
```

Comment:
Old code was fine. Use wallThickness not -wallOutwardDisplacement which was reversing colon d3.